Add PhpdocToCommentFixer

ceeram opened the pull request:
> MARK FOR MERGE INTO 2.0 by @keradus
> Remove `$token->getLine()` !

stof commented:
This forbids using inline phpdoc to document local variables

stof commented on Symfony/CS/Fixer/Contrib/PhpdocToCommentFixer.php:
```diff
@@ -0,0 +1,125 @@
+                continue;
+            }
+
+            $token->setContent('/*'.ltrim($token->getContent(), '/*'));
```
this is wrong. The token will still be marked as a T_DOC_COMMENT, which might break things (unlikely here because each iteration only looks at later parts of the collection and the collection is dropped after that, but totally sure in the 2.0 version where the Tokens collection is not recreated for each fixer).

You should be using ``$token->override`` instead

ceeram replied:
ah right, will do

stof commented on Symfony/CS/Fixer/Contrib/PhpdocToCommentFixer.php:
```diff
@@ -0,0 +1,125 @@
+            return true;
+        }
+
+        return false;
```
this can be simplified: ``return $token->isClassy() || $token->isGivenKind($skip);``

ceeram replied:
will do

stof commented on Symfony/CS/Fixer/Contrib/PhpdocToCommentFixer.php:
```diff
@@ -0,0 +1,125 @@
+    private function isValidForeach($tokens, $index)
+    {
+        $startIndex = $index;
+        $endIndex = $tokens->getNextTokenOfKind($startIndex, array(')'));
```
this is wrong. It won't work for parenthesis inside the foreach loop (array definitions, method calls, ``list`` usage, ...)

ceeram replied:
What is wrong exactly? same thing could be done for list() as well, but not for method calls.

stof replied:
the issue is that the first closing parenthesis is **not** the closing parenthesis of the foreach loop. It can be anything before that:

```php
foreach ($foo->getPairs() as $index => list($a, $b)) [
    // Do something with $index, $a and $b
}
```

ceeram replied:
ow missed this comment, will add test and fix for this.

stof commented on Symfony/CS/Tests/Fixer/Contrib/PhpdocToCommentFixerTest.php:
```diff
@@ -0,0 +1,193 @@
+ * Do not convert this
+ */
+class DocBlocks
+{
```
please add a property definition too

ceeram replied:
arent they already there?

stof replied:
sorry, I missing it

stof commented on Symfony/CS/Tests/Fixer/Contrib/PhpdocToCommentFixerTest.php:
```diff
@@ -0,0 +1,193 @@
+trait DocBlocks
+{
+    public function test() {}
+}',
```
this case should be added only on PHP 5.4+

ceeram replied:
done

stof commented on Symfony/CS/Tests/Fixer/Contrib/PhpdocToCommentFixerTest.php:
```diff
@@ -0,0 +1,193 @@
+
+/**
+ * Do not convert this
+ */
```
why here ?

ceeram replied:
why here what?

stof replied:
hmm, phpdoc indeed recommends putting a phpdoc on include statements. I'm not sure why though

ceeram replied:
i also dont know why, i used http://www.phpdoc.org/docs/latest/glossary.html#term-structural-elements as reference here.

stof commented on Symfony/CS/Fixer/Contrib/PhpdocToCommentFixer.php:
```diff
@@ -0,0 +1,125 @@
+     * Check if token is a structural element
+     * @see http://www.phpdoc.org/docs/latest/glossary.html#term-structural-elements
+     *
+     * @param $token
```
invalid phpdoc. missing type

stof commented on Symfony/CS/Fixer/Contrib/PhpdocToCommentFixer.php:
```diff
@@ -0,0 +1,125 @@
+     * @param $token
+     * @return bool
+     */
+    private function isStructuralElement($token)
```
please typehint the argument

stof commented on Symfony/CS/Fixer/Contrib/PhpdocToCommentFixer.php:
```diff
@@ -0,0 +1,125 @@
+     * @param $index
+     * @return bool
+     */
+    private function isValidForeach($tokens, $index)
```
please typehint the argument

stof commented on Symfony/CS/Fixer/Contrib/PhpdocToCommentFixer.php:
```diff
@@ -0,0 +1,125 @@
+     * Checks foreach statements for correct docblock usage.
+     *
+     * @param $tokens
+     * @param $index
```
missing param types

stof commented on Symfony/CS/Fixer/Contrib/PhpdocToCommentFixer.php:
```diff
@@ -0,0 +1,122 @@
+    /**
+     * Checks foreach statements for correct docblock usage.
+     *
+     * @param \Symfony\CS\Tokenizer\Tokens $tokens
```
this should use the short class name

stof commented on Symfony/CS/Fixer/Contrib/PhpdocToCommentFixer.php:
```diff
@@ -0,0 +1,122 @@
+     *
+     * @param \Symfony\CS\Tokenizer\Tokens $tokens
+     * @param int $index
+     * @return bool
```
please keep an empty phpdoc line between params and returns (the automatic fixer for this is not merged yet). This will also fix the test error about param alignment (currently, the are in the same group)

stof commented on Symfony/CS/Fixer/Contrib/PhpdocToCommentFixer.php:
```diff
@@ -0,0 +1,122 @@
+             * Is the $value documented in docblock: foreach($array as $value)
+             * or
+             * Is the $key documented in docblock: foreach($array as $key => $value)
+            */
```
wrong indentation

stof commented:
Please add a testcase for phpdoc for variables:

```php
/** @var int $index */
$index = $a['index'];
```

the phpdoc should not be removed in such case, and your code removes it

ceeram commented:
I have addressed all comments so far, only thing left is to allow variable assignments through `list()` calls. Will add another time, got to run now.

What about the name for this fixer? Just created one without putting too much thought into it.

stof commented:
This fixer should be configured to run before the ``phpdoc_indent`` and the ``no_empty_lines_after_phpdocs`` fixers, so that these fixers don't touch doc comments which are meant to be converted to regular comments (and the ordering should be tested in the test covering this)

ceeram commented:
How does one test ordering?

stof commented:
@ceeram https://github.com/FriendsOfPHP/PHP-CS-Fixer/blob/1.5/Symfony/CS/Tests/FixerTest.php

ceeram commented:
Thanks, added test for that as well. I think i have addressed all comments so far.

keradus commented on Symfony/CS/Fixer/Contrib/PhpdocToCommentFixer.php:
```diff
@@ -0,0 +1,187 @@
+            return false;
+        }
+
+        return strpos($tokens[$index]->getContent(), $tokens[$variable]->getContent()) !== false;
```
Yoda with you should be

ceeram replied:
ok

ceeram replied:
Imo that is taking yoda a bit too far, as there is no assignement taking place here, but will change if that makes you happy. Wasnt the Yoda fixer removed before 1.0 for the same reason?

keradus replied:
nope, please see #701.
Yoda via Symfony puts const/literal on the left side, like here.
Fixer that was temporary reverted was to move all r-values on the left, even functon calls (like change `$foo === $bar->baz()` into `$bar->baz() === $foo`.

keradus commented on Symfony/CS/Fixer/Contrib/PhpdocToCommentFixer.php:
```diff
@@ -0,0 +1,187 @@
+     */
+    private function isStructuralElement(Token $token)
+    {
+        $skip = array(
```
I would made it static.

ceeram replied:
ok

keradus commented on Symfony/CS/Fixer/Contrib/PhpdocToCommentFixer.php:
```diff
@@ -0,0 +1,187 @@
+     */
+    public function getPriority()
+    {
+        // should be run before the PhpdocIndentFixer, PhpdocParamsFixer and NoEmptyLinesAfterPhpdocsFixer
```
why?

ceeram replied:
See: https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/969#issuecomment-69777520

GrahamCampbell replied:
@ceeram, @keradus Means you need to actually say why in the source code. He's asked me to do the same in my phpdoc pull request. :)

ceeram replied:
ah ok, will add some more info

keradus replied:
It's just because no one will remember why it should be that way after few months ;)

keradus commented on Symfony/CS/Tests/FixerTest.php:
```diff
@@ -182,7 +182,9 @@ public function getFixersPriorityCases()
             array($fixers['standardize_not_equal'], $fixers['strict']),
             array($fixers['double_arrow_multiline_whitespaces'], $fixers['multiline_array_trailing_comma']),
             array($fixers['double_arrow_multiline_whitespaces'], $fixers['align_double_arrow']),
-            array($fixers['phpdoc_indent'], $fixers['phpdoc_params']),
```
why ?

ceeram replied:
oops, mistake

keradus commented on Symfony/CS/Fixer/Contrib/PhpdocToCommentFixer.php:
```diff
@@ -0,0 +1,187 @@
+
+            // skip if there is no next token or if next token is block end `}`
+            if (null === $nextIndex || $tokens[$nextIndex]->equals('}')) {
+                continue;
```
This line is not covered by tests.

keradus commented on Symfony/CS/Fixer/Contrib/PhpdocToCommentFixer.php:
```diff
@@ -0,0 +1,187 @@
+            }
+        }
+
+        return false;
```
This line is not covered by tests

ceeram commented:
All comments by @keradus have been addressed

keradus commented on Symfony/CS/Fixer/Contrib/PhpdocToCommentFixer.php:
```diff
@@ -0,0 +1,191 @@
+        $tokens = Tokens::fromCode($content);
+
+        foreach ($tokens->findGivenKind(T_DOC_COMMENT) as $index => $token) {
+            $nextIndex = $tokens->getNextMeaningfulToken($index);
```
There may be no tokens after T_DOC_COMMENT.

Instead of removing `if` statement I wasn't covered by tests here (https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/969#discussion_r22883661) please add a test for it (and recover that `if`)

ceeram replied:
In that case, docblock should be changed to comment actually. will add back and add testcase

keradus replied:
one way or another, the current form will blow up on using non-existig index ;)

ceeram replied:
see https://github.com/FriendsOfPHP/PHP-CS-Fixer/commit/a382378e024820d07001484cf0127ba29bde193c

keradus replied:
nice ;)

keradus commented on Symfony/CS/Fixer/Contrib/PhpdocToCommentFixer.php:
```diff
@@ -0,0 +1,196 @@
+    {
+        /*
+         * Should be run before the PhpdocIndentFixer, PhpdocParamsFixer and NoEmptyLinesAfterPhpdocsFixer
+         * so that these fixers don't touch doc comments which are meant to be converted to regular comments
```
So, is this mean that this should be the first phpdoc fixer of all ?
Asking in #887 context.

ceeram replied:
Yes, should also run before those new fixers

keradus commented:
nice ;)

keradus commented:
:+1:

anything else to do here before merge?

ceeram commented:
Yes there still is one thing i want to change. If docblock is right after
open tag it can be a page/file-level docblock. This fixer better not change
to comment then. I have some unpushed changes for that on a different
computer.
Op 17 jan. 2015 23:51 schreef "Dariusz Rumiński" <notifications@github.com>:

> [image: :+1:]
>
> anything else to do here before merge?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/969#issuecomment-70388030>
> .
>

keradus commented:
Then - should we consider this as done ?

keradus commented:
Please see: https://github.com/ceeram/PHP-CS-Fixer/pull/5

ceeram commented:
See: https://github.com/keradus/PHP-CS-Fixer/pull/7

ceeram commented:
All squashed now, included the additional test and fix

keradus commented:
:+1:

keradus commented on Symfony/CS/Fixer/Contrib/PhpdocToCommentFixer.php:
```diff
@@ -0,0 +1,203 @@
+                continue;
+            }
+
+            if ($this->isStructuralElement($tokens[$nextIndex])) {
```
please cache `$tokens[$nextIndex]`

keradus replied:
https://github.com/ceeram/PHP-CS-Fixer/pull/6

keradus commented on Symfony/CS/Fixer/Contrib/PhpdocToCommentFixer.php:
```diff
@@ -0,0 +1,203 @@
+        $i = 0;
+        while (!$end) {
+            ++$endIndex;
+            if ($tokens[$endIndex]->equals('(')) {
```
Can't you use `Tokens::findBlockEnd` here ?

keradus replied:
https://github.com/ceeram/PHP-CS-Fixer/pull/6

keradus commented on Symfony/CS/Fixer/Contrib/PhpdocToCommentFixer.php:
```diff
@@ -0,0 +1,203 @@
+
+            if (
+                $tokens[$nextMeaningful]->isGivenkind(T_VARIABLE) &&
+                strpos($tokens[$index]->getContent(), $tokens[$nextMeaningful]->getContent()) !== false
```
Yoda with you should be.

keradus replied:
https://github.com/ceeram/PHP-CS-Fixer/pull/6

keradus replied:
https://github.com/ceeram/PHP-CS-Fixer/pull/6

keradus commented:
ping @ceeram

ceeram commented:
pong, busy atm, will get back to this later

2015-01-21 10:02 GMT+01:00 Dariusz Rumiński <notifications@github.com>:

> ping @ceeram <https://github.com/ceeram>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/969#issuecomment-70804570>
> .
>

keradus commented:
well... maybe it should be at symfony level?

cc @stof @sstok

keradus commented:
IMO after cherry-pick from my PR it will be ready to merge.

keradus commented:
ping @ceeram 
btw, please move it into Symfony level (https://github.com/symfony/symfony/pull/13469)

keradus commented:
Thank you @ceeram.

Merged manually in #997
